File: src/medium/49-group-anagrams/python/main.py
from collections import defaultdict
from typing import List

ord_a = ord('a')
alphabetLength = (ord('z') - ord_a + 1)

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        grouped_anagrams = defaultdict(list)
        for word in strs:
            frequency = [0] * alphabetLength
            for char in word:
                frequency[ord(char) - ord_a] += 1
            grouped_anagrams[tuple(frequency)].append(word)
        return list(grouped_anagrams.values())

File: src/medium/49-group-anagrams/typescript/main.ts
const ordinalOfA = 'a'.codePointAt(0)!; // 97
const lengthOfAlphabet = 'z'.codePointAt(0)! - ordinalOfA + 1; // 26

function groupAnagrams(strs: string[]): string[][] {
	const groupedAnagrams: Record<string, string[]> = {};
	for (const string_ of strs) {
		const frequency = Array.from({length: lengthOfAlphabet}, () => 0);
		for (const char of string_) {
			frequency[char.codePointAt(0)! - ordinalOfA]++;
		}

		const key = frequency.join(',');
		groupedAnagrams[key] = [...(groupedAnagrams[key] ?? []), string_];
	}

	return Object.values(groupedAnagrams);
}
